### PR TITLE
Modularized GCM driver

### PR DESCRIPTION
**Pull request by kpamnany**
# Description

Supports Baroclinic Wave and Held-Suarez problems with selectable perturbations, base states, moisture profiles and surface fluxes.

Makes it easier to add new problems.

I have

- [X] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [X] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [X] There are no open pull requests for this already
- [X] CLIMA developers with relevant expertise have been assigned to review this submission
- [X] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: experiments/AtmosGCM/GCMDriver/GCMDriver.jl [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    _R_d::FT = R_d(bl.param_set)
+    _Ω::FT = Omega(bl.param_set)
+    _a::FT = planet_radius(bl.param_set)
+    _p_0::FT = MSLP(bl.param_set)
+    M_v::FT = 0.608
+
+    # grid
+    φ = latitude(bl.orientation, aux)
+    λ = longitude(bl.orientation, aux)
+    z = altitude(bl.orientation, bl.param_set, aux)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/GCMDriver.jl:66]:
```suggestion
    φ = latitude(bl, aux)
    λ = longitude(bl, aux)
    z = altitude(bl, aux)
```

File: experiments/AtmosGCM/GCMDriver/GCMDriver.jl [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    return nothing
+end
+
+function lowercasearg(arg)
+    if arg === nothing
+        return nothing
+    else
+        return lowercase(arg)
+    end
+end
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/GCMDriver.jl:127]:
```suggestion
lowercasearg(arg::Nothing) = nothing
lowercasearg(arg) = lowercase(arg)
```

File: experiments/AtmosGCM/GCMDriver/GCMDriver.jl [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+function config_gcm_experiment(
+    FT,
+    poly_order,
+    resolution,
+    experiment_arg,
+    wind_perturbation_arg,
+    base_state_arg,
+    moisture_profile_arg,
+    surface_flux_arg,
+)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/GCMDriver.jl:152]:
```suggestion
function config_gcm_experiment(
    ::Type{FT},
    poly_order,
    resolution,
    experiment_arg,
    wind_perturbation_arg,
    base_state_arg,
    moisture_profile_arg,
    surface_flux_arg,
) where {FT}
```

File: experiments/AtmosGCM/GCMDriver/GCMDriver.jl [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    )
+
+    return config
+end
+
+# default diagnostics configuration setup
+function config_diagnostics(FT, driver_config)
+    interval = "40000steps"
+
+    _planet_radius = planet_radius(param_set)::FT
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/GCMDriver.jl:229]: What's the difference between this and
```julia
    _planet_radius::FT = planet_radius(param_set)
```
?

File: experiments/AtmosGCM/GCMDriver/GCMDriver.jl [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        )
+        nothing
+    end
+
+    # Run the model
+    result = ClimateMachine.invoke!(
+        solver_config;
+        diagnostics_config = dgn_config,
+        user_callbacks = (cbfilter,),
+        #user_callbacks = (cbtmarfilter, cbfilter),
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/GCMDriver.jl:378]: Do we want to include the `cbtmarfilter` filter?

File: experiments/AtmosGCM/GCMDriver/gcm_bcs.jl [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# Boundary conditions
+
+# parse `--surface-flux` command line argument
+function parse_surface_flux_arg(arg, FT, param_set, orientation, moisture)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/gcm_bcs.jl:4]:
```suggestion
function parse_surface_flux_arg(arg, ::Type{FT}, param_set, orientation, moisture) where {FT}
```

File: experiments/AtmosGCM/GCMDriver/gcm_base_states.jl [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    _grav = grav(bl.param_set)::FT
+    _R_d = R_d(bl.param_set)::FT
+    _Ω = Omega(bl.param_set)::FT
+    _a = planet_radius(bl.param_set)::FT
+    _p_0 = MSLP(bl.param_set)::FT
+    M_v = 0.608::FT # TODO
+
+    # grid
+    φ = latitude(bl.orientation, aux)
+    z = altitude(bl.orientation, bl.param_set, aux)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/gcm_base_states.jl:42]:
```suggestion
    φ = latitude(bl, aux)
    z = altitude(bl, aux)
```

File: experiments/AtmosGCM/GCMDriver/gcm_bcs.jl [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+T_sfc_pole = SST at the poles (default: 271 K), specified above
+"""
+struct Varying_SST_TJ16{PS, O, MM}
+    param_set::PS
+    orientation::O
+    moisture::MM
+end
+function (st::Varying_SST_TJ16)(state, aux, t)
+    FT = eltype(state)
+    φ = latitude(st.orientation, aux)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/gcm_bcs.jl:48]: `st` doesn't seem to be defined in this function

File: experiments/AtmosGCM/GCMDriver/gcm_moisture_profiles.jl [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    aux,
+    coords,
+    t,
+    p,
+)
+    FT = eltype(state)
+
+    _p_0 = MSLP(bl.param_set)::FT
+
+    φ = latitude(bl.orientation, aux)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/gcm_moisture_profiles.jl:44]:
```suggestion
    φ = latitude(bl, aux)
```

File: experiments/AtmosGCM/GCMDriver/heldsuarez_problem.jl [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    k_f = FT(1 / _day)
+    k_s = FT(1 / (4 * _day))
+    ΔT_y = FT(60)
+    Δθ_z = FT(10)
+    T_equator = FT(315)
+    T_min = FT(200)
+    σ_b = FT(7 / 10)
+
+    # Held-Suarez forcing
+    φ = latitude(bl.orientation, aux)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/heldsuarez_problem.jl:71]:
```suggestion
    φ = latitude(bl, aux)
```

File: experiments/AtmosGCM/GCMDriver/heldsuarez_problem.jl [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    # Parameters
+    T_ref = FT(255)
+
+    # Extract the state
+    ρ = state.ρ
+    ρu = state.ρu
+    ρe = state.ρe
+
+    e_int = internal_energy(bl.moisture, bl.orientation, state, aux)
+    T = air_temperature(bl.param_set, e_int)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/heldsuarez_problem.jl:52]:
```suggestion
    ts = thermo_state(bl, state, aux)
    e_int = internal_energy(ts)
    T = air_temperature(ts)
```

File: experiments/AtmosGCM/GCMDriver/heldsuarez_problem.jl [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    k_s = FT(1 / (4 * _day))
+    ΔT_y = FT(60)
+    Δθ_z = FT(10)
+    T_equator = FT(315)
+    T_min = FT(200)
+    σ_b = FT(7 / 10)
+
+    # Held-Suarez forcing
+    φ = latitude(bl.orientation, aux)
+    p = air_pressure(bl.param_set, T, ρ)
```
Comment (charleskawczynski) [inline experiments/AtmosGCM/GCMDriver/heldsuarez_problem.jl:72]:
```suggestion
    p = air_pressure(ts)
```

File: experiments/AtmosGCM/GCMDriver/GCMDriver.jl (hunk 4 above)
Comment (kpamnany) [reply experiments/AtmosGCM/GCMDriver/GCMDriver.jl:229]: See [here](https://github.com/CliMA/ClimateMachine.jl/pull/1447#discussion_r474313647); I haven't gotten to the bottom of it yet. Will be exploring that soon.

File: experiments/AtmosGCM/GCMDriver/gcm_bcs.jl (hunk 8 above)
Comment (kpamnany) [reply experiments/AtmosGCM/GCMDriver/gcm_bcs.jl:48]: `st` is the identifier for the type -- see line 46. This is a method definition for a callable type.

**Pull request merged**